\documentclass[a4paper]{book}
\usepackage[times,hyper]{Rd}
\usepackage{makeidx}
\usepackage[latin1]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge \R{} documentation}} \par\bigskip{{\Large of \file{man/barErrorPlot.Rd} etc.}}
\par\bigskip{\large \today}
\end{center}
\inputencoding{utf8}
\HeaderA{barErrorPlot}{Generate bar error plot and its dispersion by cell types or by number of different cell types in test bulk samples.}{barErrorPlot}
%
\begin{Description}\relax
Generate bar error plot and its dispersion by cell types (\code{CellType}) or
by number of different cell types (\code{nMix}) in test bulk samples.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
barErrorPlot(
  object,
  error,
  by,
  dispersion = "se",
  filter.sc = TRUE,
  title = NULL,
  angle = 90,
  theme = theme_grey()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{DigitalDLSorter} object with \code{trained.model} slot
containing metrics in \code{eval.stats.samples} slot.

\item[\code{error}] MAE or MSE. By default it is used a list of custom colors
provided by the package.

\item[\code{by}] Variable used to display errors.

\item[\code{dispersion}] Standard error (\code{'se'}) or standard deviation
(\code{'sd'}). The first by default.

\item[\code{filter.sc}] Boolean indicating if filter single-cell profiles and only
display correlations of results associated with bulk samples (\code{TRUE}
by default).

\item[\code{title}] Title of the plot.

\item[\code{angle}] Angle of ticks.

\item[\code{theme}] ggplot theme.
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}} \code{\LinkA{corrExpPredPlot}{corrExpPredPlot}}
\code{\LinkA{distErrorPlot}{distErrorPlot}} \code{\LinkA{blandAltmanLehPlot}{blandAltmanLehPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
barErrorPlot(
  object = DDLSChung,
  error = "MSE",
  by = "CellType"
)

barErrorPlot(
  object = DDLSChung,
  error = "MAE",
  by = "nMix"
)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{barPlotCellTypes}{Plot a bar plot with deconvoluted cell type proportions.}{barPlotCellTypes}
\aliasA{barPlotCellTypes,ANY-method}{barPlotCellTypes}{barPlotCellTypes,ANY.Rdash.method}
\aliasA{barPlotCellTypes,DigitalDLSorter-method}{barPlotCellTypes}{barPlotCellTypes,DigitalDLSorter.Rdash.method}
%
\begin{Description}\relax
This function allows to plot a bar plot with the deconvoluted cell type
proportions of a given bulk RNA-seq sample using ggplot2.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
barPlotCellTypes(
  data,
  colors,
  simplify = NULL,
  color.line = NA,
  x.label = "Bulk samples",
  rm.x.text = FALSE,
  title = "Results of deconvolution",
  legend.title = "Cell types",
  angle = 90,
  ...
)

## S4 method for signature 'DigitalDLSorter'
barPlotCellTypes(
  data,
  colors = NULL,
  name.data = NULL,
  simplify = NULL,
  color.line = NA,
  x.label = "Bulk samples",
  rm.x.text = FALSE,
  title = "Results of deconvolution",
  legend.title = "Cell types",
  angle = 90
)

## S4 method for signature 'ANY'
barPlotCellTypes(
  data,
  colors,
  color.line = NA,
  x.label = "Bulk samples",
  rm.x.text = FALSE,
  title = "Results of deconvolution",
  legend.title = "Cell types",
  angle = 90
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] \code{DigitalDLSorter} object with \code{deconv.results} slot or
\code{data.frame}/\code{matrix} with cell types as columns and samples as
rows.

\item[\code{colors}] Vector with colors that will be used.

\item[\code{color.line}] Color of border bars.

\item[\code{x.label}] Label of x axis.

\item[\code{rm.x.text}] Logical value indicating if remove x axis ticks (names of
samples).

\item[\code{title}] Title of plot.

\item[\code{legend.title}] Title of legend plot.

\item[\code{angle}] Angle of text ticks.

\item[\code{name.data}] If a DigitalDLSorter is given, name of the element that
stores the results in \code{deconv.results} slot. If not, forget it.

\item[\code{simplified}] Vector with cell types that will be compressed into the
cell type with more probability in each sample by majority voting. This
option is intended for cases with exclusive cell types that they have not
sense that appear in the same sample.
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{deconvDigitalDLSorter}{deconvDigitalDLSorter}}
\code{\LinkA{deconvDigitalDLSorterObj}{deconvDigitalDLSorterObj}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
## Using a matrix
barPlotCellTypes(deconvResults)

## Using a DigitalDLSorter object
barPlotCellTypes(DDLS.Chung)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{blandAltmanLehPlot}{Generate Bland-Altman agreement plot between predicted and expected cell type proportions from test samples.}{blandAltmanLehPlot}
%
\begin{Description}\relax
Generate  Bland-Altman agreement plot between predicted and expected cell
type proportions from test samples. The  Bland-Altman agreement plots can be
displayed all mixed or split based on cell type (\code{CellType}) or the
number of cell types present in the sample (\code{nMix}). See \code{facet.by}
argument and examples for more information.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
blandAltmanLehPlot(
  object,
  colors,
  color.by,
  facet.by = NULL,
  log.2 = FALSE,
  filter.sc = TRUE,
  density = TRUE,
  color.density = "darkblue",
  size.point = 0.05,
  alpha.point = 1,
  ncol = NULL,
  nrow = NULL,
  title = NULL,
  theme = theme_grey(),
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{DigitalDLSorter} object with \code{trained.model} slot
containing metrics in \code{eval.stats.samples} slot.

\item[\code{colors}] Vector of colors to use. Only vectors with a number of colors
equal to or greater than the levels of \code{color.by} will be accepted. By
default it is used a list of custom colors provided by the package.

\item[\code{color.by}] Variable used to color data. The options are \code{nMix} and
\code{CellType}.

\item[\code{facet.by}] Variable used to display data in different panels. If it is
\code{NULL}, the plot is not separated into different panels. The options
are \code{nMix} (by number of different cell types) and \code{CellType} (by
cell type).

\item[\code{filter.sc}] Boolean indicating if filter single-cell profiles and only
display correlations of results associated with bulk samples (\code{TRUE}
by default).

\item[\code{density}] Boolean indicating if show density lines (\code{TRUE} by
default).

\item[\code{color.density}] Color of density lines if \code{density} argument is
equal to \code{TRUE}.

\item[\code{size.point}] Size of points (0.1 by default).

\item[\code{alpha.point}] Alpha of points (0.1 by default).

\item[\code{ncol}] Number of columns if \code{facet.by} is different than
\code{NULL}.

\item[\code{nrow}] Number of rows if \code{facet.by} is different than \code{NULL}.

\item[\code{title}] Title of the plot.

\item[\code{theme}] ggplot theme.

\item[\code{...}] Additional argument for \code{facet\_wrap} ggplot function if
\code{facet.by} is not equal to \code{NULL}.

\item[\code{log2}] If show  Bland-Altman agreement plot in log2 space (\code{FALSE}
by default).
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}} \code{\LinkA{corrExpPredPlot}{corrExpPredPlot}}
\code{\LinkA{distErrorPlot}{distErrorPlot}} \code{\LinkA{barErrorPlot}{barErrorPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
## Bland-Altman plot by cell type
blandAltmanLehPlot(
  object = DDLSChung,
  facet.by = "CellType",
  color.by = "CellType",
  corr = "both"
)
## Bland-Altman plot of all samples mixed
blandAltmanLehPlot(
  object = DDLSChung,
  facet.by = NULL,
  color.by = "CellType",
  alpha.point = 0.3
)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{breast.chung.generic}{Pre-trained DigitalDLSorter DNN model for deconvolution of TILs present in breast cancer environment (generic version).}{breast.chung.generic}
\keyword{datasets}{breast.chung.generic}
%
\begin{Description}\relax
DigitalDLSorter DNN model built and trained with single-cell data from Chung
et al., 2017 (GSE75688). This model allows the enumeration and quantification
of immune infiltrated cell types in breast cancer environment. This data set
consists in single-cell profiles from 11 patients from different tumor
etiology and stages (see Torroja and Sanchez-Cabo, 2019 for more details).
The analysis and characterization of the cells was carried out by the authors
of \code{digitalDLSorteR} package.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
breast.chung.generic
\end{verbatim}
\end{Usage}
%
\begin{Format}
A \code{DigitalDLSorterDNN} object with the following slots:
\begin{description}
 \item[model] Trained DNN model.
\item[training.history] Evolution of metrics and loss function during
training.\item[eval.stats] Metrics and loss results on test data.
\item[predict.results] Predictions of cell types on test data.
\item[cell.types] Cell types considered by DNN model.
\item[features] Features (genes) considered by model.
\end{description}

\end{Format}
%
\begin{Details}\relax
The cell types considered in this model are 7. They are the generic groups
from cell types considered in specific version: B cells, T CD4+ cells, T CD8+
cells, monocytes, dendritic cells, stromal cells and tumor cells.

The genes considered are 23.260 in SYMBOL notation.

The model consists in 2 hidden layers with 200 neurons per layer trained with
'kullback\_leibler\_divergence' loss function  batch size equal to 128 and a
number of epochs equal to 25.
\end{Details}
%
\begin{Source}\relax
\url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688}
\end{Source}
%
\begin{References}\relax
Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim,
K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and
immune cell profiling in primary breast cancer. Nat. Commun. 8 (1), 15081.
doi: \url{10.1038/ncomms15081}.

Torroja, C. y Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning
algorithm to quantify immune cell populations based on scRNA-Seq data.
Frontiers in Genetics 10, 978. doi: \url{10.3389/fgene.2019.00978}
\end{References}
\inputencoding{utf8}
\HeaderA{breast.chung.specific}{Pre-trained DigitalDLSorter DNN model for deconvolution of TILs present in breast cancer environment (specific version).}{breast.chung.specific}
\keyword{datasets}{breast.chung.specific}
%
\begin{Description}\relax
DigitalDLSorter DNN model built and trained with single-cell data from Chung
et al., 2017 (GSE75688). This model allows the enumeration and quantification
of immune infiltrated cell types in breast cancer environment. This data set
consists in single-cell profiles from 11 patients from different tumor
etiology and stages (see Torroja and Sanchez-Cabo, 2019 for more details).
The analysis and characterization of cells was carried out by the authors of
\code{digitalDLSorteR} package.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
breast.chung.specific
\end{verbatim}
\end{Usage}
%
\begin{Format}
A \code{DigitalDLSorterDNN} object with the following slots:
\begin{description}
 \item[model] Trained DNN model.
\item[training.history] Evolution of metrics and loss function during
training.\item[eval.stats] Metrics and loss results on test data.
\item[predict.results] Predictions of cell types on test data.
\item[cell.types] Cell types considered by DNN model.
\item[features] Features (genes) considered by model.
\end{description}

\end{Format}
%
\begin{Details}\relax
The cell types considered in this model are 13, four of them being the
intrinsic molecular subtypes of breast cancer: ER+, HER2+, ER+/HER2+, TNBC,
Stromal, Monocyte, TCD4mem (memory CD4+ T cells), BGC (germinal center B
cells), Bmem (memory B cells), DC (dendritic cells), Macrophage, TCD8 (CD8+ T
cells) and TCD4reg (regulatory CD4+ T cells).

The genes considered are 23.260 in Symbol notation.

The model consists in 2 hidden layers with 200 neurons per layer trained with
'kullback\_leibler\_divergence' loss function  batch size equal to 128 and a
number of epochs equal to 25.
\end{Details}
%
\begin{Source}\relax
\url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688}
\end{Source}
%
\begin{References}\relax
Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim,
K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and
immune cell profiling in primary breast cancer. Nat. Commun. 8 (1), 15081.
doi: \url{10.1038/ncomms15081}.

Torroja, C. y Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning
algorithm to quantify immune cell populations based on scRNA-Seq data.
Frontiers in Genetics 10, 978. doi: \url{10.3389/fgene.2019.00978}
\end{References}
\inputencoding{utf8}
\HeaderA{bulk.sim}{Get and set \code{bulk.sim} slot in a \code{DigitalDLSorter} object.}{bulk.sim}
\aliasA{bulk.sim<\Rdash{}}{bulk.sim}{bulk.sim<.Rdash.}
%
\begin{Description}\relax
Get and set \code{bulk.sim} slot in a \code{DigitalDLSorter}
object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
bulk.sim(object, type.data = "both")

bulk.sim(object, type.data = "both") <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorter} object.

\item[\code{type.data}] Element of the list. Can be 'train', 'test' or 'both' (the
last by default).

\item[\code{value}] A \code{list} with two elements, train and test, each one being
a \code{SummarizedExperiment} object with simulated bulk RNA-Seq samples.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{calculateEvalMetrics}{Calculate evaluation metrics for bulk RNA-seq samples from test data.}{calculateEvalMetrics}
%
\begin{Description}\relax
Calculate evaluation metrics for bulk RNA-seq samples from test data in order
to know the performance of the model. By default, absolute error (AbsErr),
proportional absolute error (ppAbsErr), squared error (SqrErr) and
proportional squared error (ppSqrErr) are calculated for each test sample.
Moreover, each one of these metrics are aggregated using their mean values by
three criteria: each cell type (\code{CellType}), probability bins of 0.1
(\code{pBin}), number of different cell types present in the sample
\code{nMix} and a combination of \code{pBin} and \code{nMix}
(\code{pBinNMix}). Finally, the process is repeated only for bulk samples,
removing single-cell profiles from the evaluation. Evaluation metrics are
available in \code{eval.stats.samples} slot of \code{DigitalDLSorterDNN}
object (\code{trained.model} of \code{DigitalDLSorter} object).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculateEvalMetrics(object, metrics = c("MAE", "MSE"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{DigitalDLSorter} object with \code{single.cell.final} and
\code{DigitalDLSorterDNN} slots.

\item[\code{metrics}] Metrics used for evaluating the performance of the model. Mean
absolute error (MAE) and mean squared error (MSE) by default.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A \code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} object with \code{trained.model} slot
containing a \code{DigitalDLSorterDNN} object with
\code{eval.stats.samples} slot.
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{distErrorPlot}{distErrorPlot}} \code{\LinkA{corrExpPredPlot}{corrExpPredPlot}}
\code{\LinkA{blandAltmanLehPlot}{blandAltmanLehPlot}} \code{\LinkA{barErrorPlot}{barErrorPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
DDLSChung <- calculateEvalMetrics(
  object = DDLSChung
)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cell.types}{Get and set \code{cell.types} slot in a \code{DigitalDLSorterDNN} object.}{cell.types}
\aliasA{cell.types<\Rdash{}}{cell.types}{cell.types<.Rdash.}
%
\begin{Description}\relax
Get and set \code{cell.types} slot in a \code{DigitalDLSorterDNN}
object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cell.types(object)

cell.types(object) <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorterDNN} object.

\item[\code{value}] A \code{vector} with cell types considered by DNN model.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{corrExpPredPlot}{Generate correlation plot between predicted and expected cell type proportions from test samples.}{corrExpPredPlot}
%
\begin{Description}\relax
Generate correlation plot between predicted and expected cell type
proportions from test samples. The correlation plots can be displayed all
mixed or split based on cell type (\code{CellType}) or the number of cell
types present in the sample (\code{nMix}). See \code{facet.by} argument and
examples for more information. Moreover, a correlation value selected by user
is displayed as annotation on the plots. See \code{corr} argument for
details.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
corrExpPredPlot(
  object,
  colors,
  facet.by = NULL,
  color.by = "CellType",
  corr = "both",
  filter.sc = TRUE,
  pos.x.label = 0.01,
  pos.y.label = 0.95,
  sep.labels = 0.15,
  size.point = 0.1,
  alpha.point = 1,
  ncol = NULL,
  nrow = NULL,
  title = NULL,
  theme = theme_grey(),
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{DigitalDLSorter} object with \code{trained.model} slot
containing metrics in \code{eval.stats.samples} slot.

\item[\code{colors}] Vector of colors to use. Only vectors with a number of colors
equal to or greater than the levels of \code{color.by} will be accepted. By
default it is used a list of custom colors provided by the package.

\item[\code{facet.by}] Variable used to display data in different panels. If it is
\code{NULL}, the plot is not separated into different panels. The options
are \code{nMix} (by number of different cell types) and \code{CellType} (by
cell type).

\item[\code{color.by}] Variable used to color data. The options are \code{nMix} and
\code{CellType}.

\item[\code{corr}] Correlation value displayed as annotation. The available metrics
are Pearson's correlation coefficient (\code{'pearson'}) and concordance
correlation coefficient (\code{'ccc'}). The argument can be equal to
\code{'pearson'}, \code{'ccc'} or \code{'both'} (by default).

\item[\code{filter.sc}] Boolean indicating if filter single-cell profiles and only
display correlations of results associated with bulk samples (\code{TRUE}
by default).

\item[\code{pos.x.label}] Position on the X axis of the errors annotations. 0.95 by
default.

\item[\code{pos.y.label}] Position on the Y axis of the errors annotations. 0.1 by
default.

\item[\code{sep.labels}] Space separating annotations if \code{corr} is equal to
\code{'both'}. 0.15 by default.

\item[\code{size.point}] Size of points (0.1 by default).

\item[\code{alpha.point}] Alpha of points (0.1 by default).

\item[\code{ncol}] Number of columns if \code{facet.by} is different than
\code{NULL}.

\item[\code{nrow}] Number of rows if \code{facet.by} is different than \code{NULL}.

\item[\code{title}] Title of the plot.

\item[\code{theme}] ggplot theme.

\item[\code{...}] Additional argument for \code{facet\_wrap} ggplot function if
\code{facet.by} is not equal to \code{NULL}.

\item[\code{error.labels}] Boolean indicating if show average error as annotation.
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}} \code{\LinkA{distErrorPlot}{distErrorPlot}}
\code{\LinkA{blandAltmanLehPlot}{blandAltmanLehPlot}} \code{\LinkA{barErrorPlot}{barErrorPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
## correlations by cell type
corrExpPredPlot(
  object = DDLSChung,
  facet.by = "CellType",
  color.by = "CellType",
  corr = "both"
)
## correlations of all samples mixed
corrExpPredPlot(
  DDLSChung,
  facet.by = NULL,
  color.by = "CellType",
  corr = "ccc",
  pos.x.label = 0.2,
  alpha.point = 0.3
)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{DDLSChungSmall}{\code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} 'toy' object.}{DDLSChungSmall}
\keyword{datasets}{DDLSChungSmall}
%
\begin{Description}\relax
\code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} 'toy' object containing a subset from the
original data set used for generating \code{breast.chung.generic} and
\code{breast.chung.specific} models in order to show some examples in
vignette and documentation. Moreover, it contains the corresponding
\code{ZinbParams} object in \code{zinb.params} slot. Data is stored as a
\code{SingleCellExperiment} object with counts in \code{assay} slot, cells
metadata in \code{colData} slot and genes metadata in \code{rowData} slot.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
DDLSChungSmall
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{DigitalDLSorter} of length 1.
\end{Format}
%
\begin{Details}\relax
For more information about the complete data set, see
\code{breast.chung.generic} or \code{breast.chung.specific}.
\end{Details}
%
\begin{Source}\relax
\url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688}
\end{Source}
%
\begin{References}\relax
Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim,
K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and
immune cell profiling in primary breast cancer. Nat. Commun. 8 (1), 15081.
doi: \url{10.1038/ncomms15081}.

Torroja, C. y Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning
algorithm to quantify immune cell populations based on scRNA-Seq data.
Frontiers in Genetics 10, 978. doi: \url{10.3389/fgene.2019.00978}
\end{References}
\inputencoding{utf8}
\HeaderA{deconv.data}{Get and set \code{deconv.data} slot in a \code{DigitalDLSorter} object.}{deconv.data}
\aliasA{deconv.data<\Rdash{}}{deconv.data}{deconv.data<.Rdash.}
%
\begin{Description}\relax
Get and set \code{deconv.data} slot in a \code{DigitalDLSorter}
object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
deconv.data(object, name.data = NULL)

deconv.data(object, name.data = NULL) <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorter} object.

\item[\code{name.data}] Name of the data. If it is \code{NULL} (by default),
all data contained in \code{deconv.data} slot are returned.

\item[\code{value}] A \code{list} whose names are the reference of the data stored.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{deconvDigitalDLSorterObj}{Deconvolute bulk gene expression samples (bulk RNA-Seq).}{deconvDigitalDLSorterObj}
%
\begin{Description}\relax
Deconvolute bulk gene expression samples (bulk RNA-Seq) enumerating and
quantifying the proportion of cell types present in a bulk sample. This
function needs a \code{DigitalDLSorter} object with a trained DNN model
(\code{\LinkA{trained.model}{trained.model}} slot) and bulk samples for deconvoluting in
\code{deconv.data} slot.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
deconvDigitalDLSorterObj(
  object,
  name.data,
  batch.size = 128,
  normalize = TRUE,
  simplify.set = NULL,
  simplify.majority = NULL,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} object with \code{trained.data}
and \code{deconv.data} slots.

\item[\code{name.data}] Name of the data store in \code{DigitalDLSorter} object. If
it is not provided, the first data set will be used.

\item[\code{batch.size}] Number of samples per gradient update. If unspecified,
\code{batch.size} will default to 128.

\item[\code{normalize}] Normalize data before deconvolution. \code{TRUE} by default.

\item[\code{simplify.set}] List specifying which cell types should be compressed
into a new label whose name will be the list item. See examples for
details. The results are stored in a list with normal and simpli.majority
results (if provided). The name of the element in the list is
\code{'simpli.set'}.

\item[\code{simplify.majority}] List specifying which cell types should be
compressed into the cell types with greater proportion in each sample.
Unlike \code{simplify.set}, it allows to maintain the complexity of the
results while compressing the information, because it is not created a new
label. The results are stored in a list with normal and simpli.set results
(if provided). The name of the element in the list is
\code{'simpli.majority'}.

\item[\code{verbose}] Show informative messages during the execution.
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is oriented for users that have trained a DNN model using their
own data. If you want to use a pre-trained model, see
\code{\LinkA{deconvDigitalDLSorter}{deconvDigitalDLSorter}}.
\end{Details}
%
\begin{Value}
A \code{data.frame} with samples (\eqn{i}{}) as rows and cell types
(\eqn{j}{}) as columns. Each entry represents the proportion of \eqn{j}{} cell
type in \eqn{i}{} sample.
\end{Value}
%
\begin{References}\relax
Torroja, C. y Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep
Learning algorithm to quantify immune cell populations based on scRNA-Seq
data. Frontiers in Genetics 10, 978. doi: \url{10.3389/fgene.2019.00978}
\end{References}
%
\begin{SeeAlso}\relax
\code{\LinkA{trainDigitalDLSorterModel}{trainDigitalDLSorterModel}} \code{\LinkA{DigitalDLSorter}{DigitalDLSorter}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
## simplify arguments
simplify <- list(Tumor = c("ER+", "HER2+", "ER+/HER2+", "TNBC"),
                 Bcells = c("Bmem", "BGC"))

## all results are stored in DigitalDLSorter object
DDLSChung <- deconvDigitalDLSorterObj(
  object = DDLSChung,
  name.data = "TCGA.small",
  normalize = TRUE,
  simplify.set = simplify,
  simplify.majority = simplify
)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{deconvDigitalDLSorter}{Deconvolute bulk gene expression samples (bulk RNA-Seq) using a pre-trained DigitalDLSorter model.}{deconvDigitalDLSorter}
%
\begin{Description}\relax
Deconvolute bulk gene expression samples (RNA-Seq) quantifying the proportion
of cell types present in a bulk sample. See in Details the available models.
This method uses a pre-trained Deep Neural Network model to enumerate and
quantify the cell types present in bulk RNA-Seq samples. For the moment, the
available models allow to deconvolute the immune infiltration breast cancer
(Chung et al., 2017) at two levels: specific cell types
(\code{'breast.chung.specific'}) and generic cell types
(\code{'breast.chung.generic'}). See \code{\LinkA{breast.chung.generic}{breast.chung.generic}} and
\code{\LinkA{breast.chung.specific}{breast.chung.specific}} documentation for details.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
deconvDigitalDLSorter(
  data,
  model = "breast.generic",
  batch.size = 128,
  normalize = TRUE,
  simplify.set = NULL,
  simplify.majority = NULL,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A \code{matrix} or a \code{data.frame} with bulk gene expression
of samples. Rows must be genes in symbol notation and columns must be
samples.

\item[\code{model}] Pre-trained DNN model to use for deconvoluting process. For the
moment, the available models are for RNA-Seq samples from breast cancer
(\code{'breast.chung.generic'} and \code{'breast.chung.specific'})
environment.

\item[\code{batch.size}] Number of samples loadad in-memory each time of
deconvolution process. If unspecified, \code{batch.size} will default to
128.

\item[\code{normalize}] Normalize data before deconvolution. \code{TRUE} by default.

\item[\code{simplify.set}] List specifying which cell types should be compressed
into a new label whose name will be the list name item. See examples for
details.

\item[\code{simplify.majority}] List specifying which cell types should be
compressed into the cell type with greater proportions in each sample.
Unlike \code{simplify.set}, it allows to maintain the complexity of the
results while compressing the information, because it is not created a new
label.

\item[\code{verbose}] Show informative messages during the execution.
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is oriented for users that only want to use the method for
deconvoluting their bulk RNA-Seq samples. For users that are building their
own model from scRNA-seq, see \code{\LinkA{deconvDigitalDLSorterObj}{deconvDigitalDLSorterObj}}. The
former works with base classes, while the last uses \code{DigitalDLSorter}
objects.

For situations where there are cell types exclusive to each other because it
does not make sense that they appear together, see arguments
\code{simplify.set} and \code{simplify.majority}.
\end{Details}
%
\begin{Value}
A \code{data.frame} with samples (\eqn{i}{}) as rows and cell types
(\eqn{j}{}) as columns. Each entry represents the predicted proportion of
\eqn{j}{} cell type in \eqn{i}{} sample.
\end{Value}
%
\begin{References}\relax
Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim,
K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and
immune cell profiling in primary breast cancer. Nat. Commun. 8 (1), 15081.
doi: \url{10.1038/ncomms15081}.
\end{References}
%
\begin{SeeAlso}\relax
\code{\LinkA{deconvDigitalDLSorterObj}{deconvDigitalDLSorterObj}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
results1 <- deconvDigitalDLSorter(
  data = TCGA.breast.small,
  model = "breast.chung.specific",
  normalize = TRUE
)

## simplify arguments
simplify <- list(Tumor = c("ER+", "HER2+", "ER+/HER2+", "TNBC"),
                 Bcells = c("Bmem", "BGC"))

## in this case,  the item names from list will be the new labels
results2 <- deconvDigitalDLSorter(
  TCGA.breast.small,
  model = "breast.chung.specific",
  normalize = TRUE,
  simplify.set = simplify)

## in this case, the cell type with greatest proportion will be the new label
## the rest of proportion cell types will be added to the greatest
results3 <- deconvDigitalDLSorter(
  TCGA.breast.small,
  model = "breast.chung.specific",
  normalize = TRUE,
  simplify.majority = simplify)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{deconv.results}{Get and set \code{deconv.results} slot in a \code{DigitalDLSorter} object.}{deconv.results}
\aliasA{deconv.results<\Rdash{}}{deconv.results}{deconv.results<.Rdash.}
%
\begin{Description}\relax
Get and set \code{deconv.results} slot in a \code{DigitalDLSorter}
object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
deconv.results(object, name.data = NULL)

deconv.results(object, name.data = NULL) <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorter} object.

\item[\code{name.data}] Name of the data. If it is \code{NULL} (by default),
all results contained in \code{deconv.results} slot are returned.

\item[\code{value}] A \code{list} whose names are the reference of the results
stored.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{DigitalDLSorter-class}{The DigitalDLSorter Class.}{DigitalDLSorter.Rdash.class}
\aliasA{DigitalDLSorter}{DigitalDLSorter-class}{DigitalDLSorter}
%
\begin{Description}\relax
The DigitalDLSorter object is the core of digitalDLSorteR. This object stores
the different intermediate data resulting from running pipeline from real
single-cell data to the trained Deep Neural Network, including the data on
which to carry out the process of devonvolution. Only it is used in the case
of building new deconvolution models. For deconvoluting bulk samples using
pre-trained models, see \code{\LinkA{deconvDigitalDLSorter}{deconvDigitalDLSorter}} function.
\end{Description}
%
\begin{Details}\relax
This object uses other classes to store the different type of data produced
during the process: \begin{itemize}
 \item \code{SingleCellExperiment} class for
single-cell RNA-seq data, using sparse matrix from the \code{Matrix} package
(\code{dgCMatrix} class) to store the matrix of counts. \item 
\code{ZinbParams} class with the estimated parameters for the simulation of
new single-cell profiles. \item \code{SummarizedExperiment} class for storing
bulk RNA-seq data. In this case, it is possible to load all data in memory or
the use of HDF5 files as back-end by \code{DelayedArray} and \code{HDF5Array}
packages. See \code{\LinkA{generateBulkSamples}{generateBulkSamples}} for details. \item 
\code{\LinkA{ProbMatrixCellTypes}{ProbMatrixCellTypes}} class for the composition cell matrices
built during the process. See \code{?ProbMatrixCellTypes} for details. \item 
\code{\LinkA{DigitalDLSorterDNN}{DigitalDLSorterDNN}} class for storing the trained Deep Neural
Network. This step is performed by \code{keras}. See
\code{\LinkA{DigitalDLSorterDNN}{DigitalDLSorterDNN}} for details. 
\end{itemize}

\end{Details}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{single.cell.real}] Real single-cell data stored in a
\code{SingleCellExperiment} object. The counts matrix is stored as a
\code{dgCMatrix} object to optimize the amount of used memory.

\item[\code{zinb.params}] \code{ZinbParams} object with estimated parameters for the
simulation of new single-cell expression profiles.

\item[\code{single.cell.final}] Final single-cell expression profiles used for
simulating bulk RNA-seq profiles with known cell composition.

\item[\code{prob.cell.types}] \code{ProbMatrixCellTypes} class with the cell
composition  matrix built for the simulation of bulk RNA-seq profiles. The
entries determine the proportion of single-cell types that will constitute
the simulated bulk samples.

\item[\code{bulk.sim}] A list with two elements: train and test simulated bulk
RNA-seq. This data are stored as a \code{SummarizedExperiment} object. We
recommend the use of HDF5 file as a back-end due to the large amount of
memory that they occupy.

\item[\code{final.data}] The final data that will be used for training and testing
the Deep Neural Network. As in the previous slot, it is a list with two
items, train and test. With respect to train counts matrix, it can be the
train bulk RNA-seq samples, the train scRNA-seq samples or a combination of
both. In the case of test counts matrix, RNA-seq data from bulk and
single-cell will be combined. Moreover, data is scaled and shuffled for
training.

\item[\code{trained.model}] \code{\LinkA{DigitalDLSorterDNN}{DigitalDLSorterDNN}} object with the trained
model, different metrics obtained during the training and evaluation
metrics from the application of the model on test data. After executing
\code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}}, it is alto possible to find the results
of the model evaluation.

\item[\code{deconv.data}] Optional slot where is possible to store new bulk samples
for its deconvolution. It is a list whose name is the name of the data
provided. It is possible to store more than one dataset to make
predictions. See \code{\LinkA{deconvDigitalDLSorterObj}{deconvDigitalDLSorterObj}} for details.

\item[\code{deconv.results}] Slot where the results from the deconvolution process
over \code{\LinkA{deconv.data}{deconv.data}} data are stored. It is a list whose name is
the name of the data from which they come.

\item[\code{project}] Name of the project.

\item[\code{version}] Version of DigitalDLSorteR this object was built under.

The package can be used in two ways: to build new models of deconvolution
from scRNA-seq data or to deconvolute bulk RNA-seq samples using
pre-trtained models integrated into the package. If you want to build new
models, see \code{\LinkA{loadRealSCProfiles}{loadRealSCProfiles}} or
\code{\LinkA{loadFinalSCProfiles}{loadFinalSCProfiles}} functions. If yoy want to use pre-trained
models, see \code{\LinkA{deconvDigitalDLSorter}{deconvDigitalDLSorter}} function.

\end{description}
\end{Section}
\inputencoding{utf8}
\HeaderA{DigitalDLSorterDNN-class}{The DigitalDLSorterDNN Class.}{DigitalDLSorterDNN.Rdash.class}
\aliasA{DigitalDLSorterDNN}{DigitalDLSorterDNN-class}{DigitalDLSorterDNN}
%
\begin{Description}\relax
The DigitalDLSorterDNN object stores the trained Deep Neural Network, the
training history of selected metrics and the results of prediction on test
data. After executing \code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}}, it is alto possible
to find the results of the model evaluation.
\end{Description}
%
\begin{Details}\relax
The steps related with Deep Learning are carried out with \code{keras}
package, so the model are stored in a R6 class, system used by the package.
If you want to save the object in an rds file, \code{digitalDLSorteR}
provides an \code{saveRDS} generic that transforms the keras model into a
native valid R object. Specifically, the model is converted into a list with
the architecture of the network and the weights learned during the training.
The is the minimum information to use the model as predictor. If you want to
maintain the optimizer state, see \code{\LinkA{saveTrainedModelAsH5}{saveTrainedModelAsH5}}
function. If you want to store an object as rda file, see
\code{\LinkA{preparingToSave}{preparingToSave}} function.
\end{Details}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{model}] Trained Deep Neural Network model. This slot can contain a R6
\code{keras.engine.sequential.Sequential} object or a list with two
elements: the architecture of the model and the resulting weights after
training.

\item[\code{training.history}] List with the evolution of the selected metrics during
training.

\item[\code{eval.stats.model}] Performance of the model on test data.

\item[\code{predict.results}] Deconvolution results matrix of test data. Columns are
cell types, rows are samples and each entry is the proportion of this cell
type on this sample.

\item[\code{cell.types}] Vector with the cell types to deconvolute.

\item[\code{features}] Vector with features used during training. These features will
be used for the following predictions.

\item[\code{eval.stats.samples}] Performance of the model on each sample of test data
in comparison with the known cell proportions.

\end{description}
\end{Section}
\inputencoding{utf8}
\HeaderA{digitLDLSorteR}{digitalDLSorteR: R package for deconvolution of bulk RNA-Seq samples based on Deep Learning.}{digitLDLSorteR}
%
\begin{Description}\relax
\emph{digitalDLSorteR} is an R package that implements a Deep Learning based
method to enumerate and quantify the cell type composition of bulk RNA-Seq
samples. Our method makes use of Deep Neural Network (DNN) models to adjust
any cell type composition starting from single-cell RNA-Seq (scRNA-Seq) data.
\end{Description}
%
\begin{Details}\relax
The rationale of the method consists in a process that starts from scRNA-Seq
data and, after a few steps, a Deep Neural Network (DNN) model is trained
with simulated bulk RNA-seq samples whose cell composition is known. The
trained model is able to deconvolve any bulk RNA-seq sample by determining
the proportion of the different cell types present in it. The main advantage
of this method is the possibility of building deconvolution models trained
with real data which comes from certain biological environments. For example,
for quantifying the proportion of tumor infiltrated lymphocytes (TILs) in
breast cancer, by following this protocol you can obtain a specific model for
this type of samples. This fact overcomes the limitation of other methods,
since stromal and immune cells change significantly their profiles depending
on the tissue and disease context.

The package can be used in two ways: for deconvolving bulk RNA-seq samples
using a pre-trained model provided by us or for building your own models
trained from your own scRNA-seq samples. These new models may be published in
order to make them available for other users that work with similar data
(e.g. neural environment, prostate cancer environment, etc.). For the moment,
the available models allows the deconvolution of TILs from breast cancer
classified by our team.
\end{Details}
\inputencoding{utf8}
\HeaderA{distErrorPlot}{Generate box plot or violin plot showing how errors are distributed.}{distErrorPlot}
%
\begin{Description}\relax
Generate violin plot or box plot showing how errors are distributed by
proportion bins of 0.1. The errors can be displayed all mixed or split based
on cell type (\code{CellType}) or number of cell types present in the sample
(\code{nMix}). See \code{facet.by} argument and examples for more
information.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
distErrorPlot(
  object,
  error,
  colors,
  x.by = "pBin",
  facet.by = NULL,
  color.by = "nMix",
  filter.sc = TRUE,
  error.labels = FALSE,
  pos.x.label = 4.6,
  pos.y.label = NULL,
  size.point = 0.1,
  alpha.point = 1,
  type = "violinplot",
  ylimit = NULL,
  nrow = NULL,
  ncol = NULL,
  title = NULL,
  theme = theme_grey(),
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{DigitalDLSorter} object with \code{trained.model} slot
containing metrics in \code{eval.stats.samples} slot.

\item[\code{error}] Which error is going to represent. The available errors are
absolute error (\code{"AbsErr"}), proportional absolute error
(\code{"ppAbsErr"}), squared error (\code{"SqrErr"}) or proportional
squared error (\code{"ppSqrErr"}).

\item[\code{colors}] Vector of colors to use. Only vectors with a number of colors
equal to or greater than the levels of \code{color.by} will be accepted. By
default it is used a list of custom colors provided by the package.

\item[\code{x.by}] Variable used for x axis. When \code{facet.by} is not
\code{NULL}, the best option is \code{pBin} (probability bin). The options
are \code{nMix} (by number of different cell types), \code{CellType} (by
cell type) and \code{pBin}.

\item[\code{facet.by}] Variable used to display data in different panels. If it is
\code{NULL}, the plot is not separated into different panels. The options
are \code{nMix} (by number of different cell types) and \code{CellType} (by
cell type).

\item[\code{color.by}] Variable used to color data. The options are \code{nMix} and
\code{CellType}.

\item[\code{filter.sc}] Boolean indicating if filter single-cell profiles and only
display errors associated with bulk samples (\code{TRUE} by default).

\item[\code{error.labels}] Boolean indicating if show average error as annotation.

\item[\code{pos.x.label}] Position on the X axis of the errors annotations.

\item[\code{pos.y.label}] Position on the Y axis of the errors annotations.

\item[\code{size.point}] Size of points (0.1 by default).

\item[\code{alpha.point}] Alpha of points (0.1 by default).

\item[\code{type}] Type of plot, \code{'boxplot'} or \code{'violinplot'}. The last
by default.

\item[\code{ylimit}] Upper limit in y axis if it is needed. \code{NULL} by default.

\item[\code{nrow}] Number of rows if \code{facet.by} is different than \code{NULL}.

\item[\code{ncol}] Number of columns if \code{facet.by} is different than
\code{NULL}.

\item[\code{title}] Title of the plot.

\item[\code{theme}] ggplot theme.

\item[\code{...}] Additional argument for \code{facet\_wrap} ggplot function if
\code{facet.by} is not equal to \code{NULL}.
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{calculateEvalMetrics}{calculateEvalMetrics}} \code{\LinkA{corrExpPredPlot}{corrExpPredPlot}}
\code{\LinkA{blandAltmanLehPlot}{blandAltmanLehPlot}} \code{\LinkA{barErrorPlot}{barErrorPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
distErrorPlot(
  object = DDLSChung,
  error = "AbsErr",
  facet.by = "CellType",
  color.by = "nMix",
  error.labels = TRUE,
  theme = theme_bw()
)

distErrorPlot(
  object = DDLSChung,
  error = "AbsErr",
  x.by = "CellType",
  facet.by = NULL,
  filter.sc = FALSE,
  color.by = "CellType",
  error.labels = TRUE
)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{estimateZinbwaveParams}{Estimate parameters for ZINB-WaVE model for simulating new single-cell expression profiles.}{estimateZinbwaveParams}
%
\begin{Description}\relax
Estimate parameters for the ZINB-WaVE model from a real single-cell
data set using ZINB-WaVE model.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
estimateZinbwaveParams(
  object,
  cell.ID.column,
  gene.ID.column,
  cell.type.column,
  cell.cov.columns,
  gene.cov.columns,
  set.type = "All",
  threads = 1,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] \code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} object with a
\code{single.cell.real} slot.

\item[\code{cell.ID.column}] Name or number of the column in cells metadata
corresponding with cell names in expression matrix.

\item[\code{gene.ID.column}] Name or number of the column in genes metadata
corresponding with the notation used for features/genes.

\item[\code{cell.type.column}] Name or number of the column in cells metadata
corresponding with cell type of each cell.

\item[\code{cell.cov.columns}] Name or number of columns in cells metadata that
will be used as covariates in the model during the estimation.

\item[\code{gene.cov.columns}] Name or number of columns in genes metadata that will
be used as covariates in the model during estimation.

\item[\code{set.type}] Cell type to evaluate. 'All' by default.

\item[\code{threads}] Number of threads used for the estimation. For setting the
parallel environment \code{BiocParallel} package is used.

\item[\code{verbose}] Show informative messages during the execution.
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
ZINB-WaVE is a flexible model for zero-inflated count data. This function
carries out the model fit to real single-cell data modeling \eqn{Y_{ij}}{}
(the count of feature \eqn{j}{} for sample \eqn{i}{}) as a random variable
following a zero-inflated negative binomial (ZINB) distribution. The
estimated parameters will be used for the simulation of new single-cell
expression profiles by sampling a negative binomial distribution and
introducing dropouts from a binomial distribution.
To do this, \code{\LinkA{DigitalDLSorter}{DigitalDLSorter}} uses \code{zinbEstimate} function
from \code{splatter} package (Zappia et al., 2017),  that is a wrapper around
\code{zinbFit} function from \code{zinbwave} package (Risso et al., 2018).
For more details about the model, see Risso et al., 2018.
\end{Details}
%
\begin{Value}
A \code{DigitalDLSorter} object with \code{zinb.params} slot
containing
a \code{ZinbParams} object. This object contains the estimated ZINB
parameters from real single-cell data.
\end{Value}
%
\begin{References}\relax
Risso, D., Perraudeau, F., Gribkova, S. et al. (2018). A general and flexible
method for signal extraction from single-cell RNA-seq data. Nat Commun 9,
284. doi: \url{\https://doi.org/10.1038/s41467-017-02554-5}.

Torroja, C. y Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning
algorithm to quantify immune cell populations based on scRNA-Seq data.
Frontiers in Genetics 10,
978. doi: \url{10.3389/fgene.2019.00978}

Zappia, L., Phipson, B. y Oshlack, A. Splatter: simulation of single-cell RNA
sequencing data. Genome Biol. 2017; 18: 174.
\end{References}
%
\begin{SeeAlso}\relax
\code{\LinkA{simSingleCellProfiles}{simSingleCellProfiles}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
DDLSChung <- estimateZinbwaveParams(
  object = DDLSChung,
  cell.ID.column = "Cell_ID",
  gene.ID.column = "external_gene_name",
  cell.type.column = "Cell_type",
  cell.cov.columns = c("Patient", "Sample_type"),
  gene.cov.columns = "gene_length",
  verbose = TRUE
)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{eval.stats.model}{Get and set \code{eval.stats.model} slot in a \code{DigitalDLSorterDNN} object.}{eval.stats.model}
\aliasA{eval.stats.model<\Rdash{}}{eval.stats.model}{eval.stats.model<.Rdash.}
%
\begin{Description}\relax
Get and set \code{eval.stats.model} slot in a
\code{DigitalDLSorterDNN} object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
eval.stats.model(object)

eval.stats.model(object) <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorterDNN} object.

\item[\code{value}] A \code{list} object with the resulting metrics after prediction
on test data with DNN model.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{eval.stats.samples}{Get and set \code{eval.stats.samples} slot in a \code{DigitalDLSorterDNN} object.}{eval.stats.samples}
\aliasA{eval.stats.samples<\Rdash{}}{eval.stats.samples}{eval.stats.samples<.Rdash.}
%
\begin{Description}\relax
Get and set \code{eval.stats.samples} slot in a
\code{DigitalDLSorterDNN} object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
eval.stats.samples(object, metrics = "All")

eval.stats.samples(object, metrics = "All") <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorterDNN} object.

\item[\code{value}] A \code{list} with evaluation metrics used for evaluating the
performance of the model over each sample from test data.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{features}{Get and set \code{features} slot in a \code{DigitalDLSorterDNN} object.}{features}
\aliasA{features<\Rdash{}}{features}{features<.Rdash.}
%
\begin{Description}\relax
Get and set \code{features} slot in a \code{DigitalDLSorterDNN}
object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
features(object)

features(object) <- value
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] A \code{DigitalDLSorterDNN} object.

\item[\code{value}] A \code{vector} with features (genes) considered by DNN model.
\end{ldescription}
\end{Arguments}
